backend/src/email/utils/callback/update-status.ts

```diff
@@ -0,0 +1,68 @@
+import {
+  addToBlacklist,
+  updateMessageWithError,
+  updateMessageWithSuccess,
+  haltCampaignIfThresholdExceeded,
+} from './query'
+import {
+  BounceMetadata,
+  ComplaintMetadata,
+  Metadata,
+} from '@email/interfaces/callback.interface'
+
+export const updateDeliveredStatus = async (
+  metadata: Metadata
+): Promise<void> => {
+  await updateMessageWithSuccess(metadata)
+}
+/**
+ *  Updates the error_code in the email_messages table for bounced delivery.
+ *  Hard bounce: The recipient's mail server permanently rejected the email.
+ *  Soft bounce: SES fails to deliver the email after retrying for a period of time.
+ *  @param message JSON object that contains the notification details
+ *  @param timestamp ISO string from notification timestamp
+ */
+export const updateBouncedStatus = async (
+  metadata: BounceMetadata
+): Promise<void> => {
+  const bounceType = metadata.bounceType
+  const recipients = metadata.to
+  let errorCode
+
+  if (bounceType === 'Permanent') {
+    errorCode = 'Hard bounce'
+    // Add to black list
+    if (recipients) await Promise.all(recipients.map(addToBlacklist))
+  } else {
+    errorCode = 'Soft bounce'
+  }
+
+  const campaignId = await updateMessageWithError({
+    errorCode,
+    timestamp: metadata.timestamp,
+    id: metadata.id,
+  })
+  await haltCampaignIfThresholdExceeded(campaignId)
+}
+
+/**
+ * Updates the error_code in the email_messages table for complaints.
+ *  @param message JSON object that contains the notification details
+ *  @param timestamp ISO string from notification timestamp
+ */
+export const updateComplaintStatus = async (
+  metadata: ComplaintMetadata
+): Promise<void> => {
+  const errorCode = metadata.complaintType
+  const recipients = metadata.to
+
+  if (errorCode && recipients) {
+    await Promise.all(recipients.map(addToBlacklist))
```

Comment:
Sorry I can't recall whats the discussion around this. With the unsubscribe digest, is it still necessary to add the complaints to blacklist?

Reply:
You're right in a way that users might be more inclined to click unsubscribe than to mark as spam. But if the user sends an unsubscribe request, and the agency user doesn't respect it and remove that email address from the recipients list, they're still going to receive messages from Postman, which they'll continue marking as spam and affect our reputation.